### PR TITLE
copr: check outdated task

### PR DESCRIPTION
**Pull request by BusyJay**
ref #1305 

@siddontang @shenli @disksing PTAL

### CODE REVIEW

File: src/server/coprocessor/endpoint.rs [hunk 1]
```diff
@@ -167,10 +181,18 @@ fn on_error(e: Error, cb: ResponseHandler) {
...
         Error::Locked(info) => resp.set_locked(info),
         Error::Other(_) => resp.set_other_error(format!("{}", e)),
+        Error::Outdated(deadline, now, tp) => return on_req_outdated(deadline, now, tp),
     }
     cb(resp)
 }
 
+fn on_req_outdated(deadline: Instant, now: Instant, tp: i64) {
+    let t = get_req_type_str(tp);
+    OUTDATED_REQ_COUNTER.with_label_values(&["select", t]).inc();
```
Comment (siddontang) [inline src/server/coprocessor/endpoint.rs]: OUTDATED_REQ_WAIT_TIME has already count, we don't need OUTDATED_REQ_COUNTER

**Comment by siddontang** [comment]
Any test to cover outdated?

File: src/server/coprocessor/endpoint.rs [hunk 2]
```diff
@@ -647,6 +698,9 @@ impl<'a> SelectContext<'a> {
                                                      },
                                                      self.key_only()));
             while limit > row_count {
+                if row_count & 255 == 0 {
```
Comment (BusyJay) [inline src/server/coprocessor/endpoint.rs]: To avoid frequent check.

Comment (siddontang) [inline src/server/coprocessor/endpoint.rs]: please use a const name for 255 here.

**Comment by siddontang** [comment]
LGTM

File: src/server/coprocessor/endpoint.rs [hunk 3]
```diff
@@ -87,13 +93,18 @@ impl Display for Task {
...
 pub struct RequestTask {
     req: Request,
+    // The deadline before which the task should be responded.
+    deadline: Instant,
     on_resp: OnResponse,
 }
 
 impl RequestTask {
     pub fn new(req: Request, on_resp: OnResponse) -> RequestTask {
+        let mut deadline = Instant::now();
```
Comment (ngaut) [inline src/server/coprocessor/endpoint.rs]: let  deadline = Instant::now() + Duration::from_secs(REQUEST_MAX_HANDLE_SECS);;

File: src/server/coprocessor/mod.rs [hunk 4]
```diff
@@ -34,6 +35,9 @@ quick_error! {
             description("key is locked")
             display("locked {:?}", l)
         }
+        Outdated(deadline: Instant, now: Instant, tp: i64) {
+            description("request is outdated")
```
Comment (ngaut) [inline src/server/coprocessor/mod.rs]: s/outdated/timeout/

**Comment by ngaut** [comment]
LGTM

**Pull request merged**